File: Kordia_backend/app/database/repositories/playlists.py
import aiosqlite
from typing import List, Dict, Any, Optional
from datetime import datetime

class PlaylistRepository:
    def __init__(self, db: aiosqlite.Connection):
        self.db = db

    async def get_all_playlists(self) -> List[Dict[str, Any]]:
        async with self.db.execute("SELECT id, name, cover_thumbnail, created_at FROM playlists ORDER BY created_at DESC") as cursor:
            rows = await cursor.fetchall()
            return [
                {
                    "id": row[0],
                    "name": row[1],
                    "coverThumbnail": row[2],
                    "createdAt": row[3],
                    "songs": await self.get_playlist_songs(row[0])
                }
                for row in rows
            ]

    async def get_playlist(self, playlist_id: str) -> Optional[Dict[str, Any]]:
        async with self.db.execute("SELECT id, name, cover_thumbnail, created_at FROM playlists WHERE id = ?", (playlist_id,)) as cursor:
            row = await cursor.fetchone()
            if not row:
                return None
            return {
                "id": row[0],
                "name": row[1],
                "coverThumbnail": row[2],
                "createdAt": row[3],
                "songs": await self.get_playlist_songs(playlist_id)
            }

    async def get_playlist_songs(self, playlist_id: str) -> List[Dict[str, Any]]:
        query = """
            SELECT ytid, title, artist, thumbnail 
            FROM playlist_songs 
            WHERE playlist_id = ? 
            ORDER BY song_order ASC
        """
        async with self.db.execute(query, (playlist_id,)) as cursor:
            rows = await cursor.fetchall()
            return [
                {
                    "ytid": row[0],
                    "title": row[1],
                    "artist": row[2],
                    "thumbnail": row[3]
                }
                for row in rows
            ]

    async def create_playlist(self, playlist_id: str, name: str) -> Dict[str, Any]:
        created_at = datetime.utcnow().isoformat()
        await self.db.execute(
            "INSERT INTO playlists (id, name, created_at) VALUES (?, ?, ?)",
            (playlist_id, name, created_at)
        )
        await self.db.commit()
        return await self.get_playlist(playlist_id)

    async def update_playlist_name(self, playlist_id: str, name: str) -> bool:
        cursor = await self.db.execute("UPDATE playlists SET name = ? WHERE id = ?", (name, playlist_id))
        await self.db.commit()
        return cursor.rowcount > 0

    async def delete_playlist(self, playlist_id: str) -> bool:
        cursor = await self.db.execute("DELETE FROM playlists WHERE id = ?", (playlist_id,))
        await self.db.commit()
        return cursor.rowcount > 0

    async def add_song(self, playlist_id: str, song: Dict[str, Any]) -> bool:
        async with self.db.execute("SELECT 1 FROM playlist_songs WHERE playlist_id = ? AND ytid = ?", (playlist_id, song['ytid'])) as cursor:
            if await cursor.fetchone():
                return False
                
        async with self.db.execute("SELECT MAX(song_order) FROM playlist_songs WHERE playlist_id = ?", (playlist_id,)) as cursor:
            row = await cursor.fetchone()
            next_order = (row[0] or 0) + 1
            
        await self.db.execute(
            "INSERT INTO playlist_songs (playlist_id, ytid, title, artist, thumbnail, song_order) VALUES (?, ?, ?, ?, ?, ?)",
            (playlist_id, song['ytid'], song['title'], song.get('artist', ''), song.get('thumbnail', ''), next_order)
        )
        
        async with self.db.execute("SELECT cover_thumbnail FROM playlists WHERE id = ?", (playlist_id,)) as cursor:
            pl_row = await cursor.fetchone()
            if pl_row and not pl_row[0] and song.get('thumbnail'):
                await self.db.execute("UPDATE playlists SET cover_thumbnail = ? WHERE id = ?", (song['thumbnail'], playlist_id))
                
        await self.db.commit()
        return True

    async def remove_song(self, playlist_id: str, ytid: str) -> bool:
        cursor = await self.db.execute("DELETE FROM playlist_songs WHERE playlist_id = ? AND ytid = ?", (playlist_id, ytid))
        
        # Verify if needs a new cover
        if cursor.rowcount > 0:
            songs = await self.get_playlist_songs(playlist_id)
            new_cover = songs[0]['thumbnail'] if songs and songs[0]['thumbnail'] else None
            await self.db.execute("UPDATE playlists SET cover_thumbnail = ? WHERE id = ?", (new_cover, playlist_id))
            
        await self.db.commit()
        return cursor.rowcount > 0

File: Kordia_backend/src/database/playlists.ts
import type Database from 'better-sqlite3';

interface PlaylistRow { id: string; name: string; cover_thumbnail: string | null; created_at: string; }
interface SongRow { ytid: string; title: string; artist: string | null; thumbnail: string | null; }

interface PlaylistDict { id: string; name: string; coverThumbnail: string | null; createdAt: string; songs: SongDict[]; }
interface SongDict { ytid: string; title: string; artist: string | null; thumbnail: string | null; }

export class PlaylistRepository {
  constructor(private db: Database.Database) {}

  getAllPlaylists(): PlaylistDict[] {
    const playlists = this.db.prepare('SELECT id, name, cover_thumbnail, created_at FROM playlists ORDER BY created_at DESC').all() as PlaylistRow[];
    return playlists.map(p => ({ ...this.toPlaylistDict(p), songs: this.getPlaylistSongs(p.id) }));
  }

  getPlaylist(id: string): PlaylistDict | undefined {
    const row = this.db.prepare('SELECT id, name, cover_thumbnail, created_at FROM playlists WHERE id = ?').get(id) as PlaylistRow | undefined;
    if (!row) return undefined;
    return { ...this.toPlaylistDict(row), songs: this.getPlaylistSongs(id) };
  }

  getPlaylistSongs(playlistId: string): SongDict[] {
    return this.db.prepare(
      'SELECT ytid, title, artist, thumbnail FROM playlist_songs WHERE playlist_id = ? ORDER BY song_order ASC'
    ).all(playlistId) as SongDict[];
  }

  createPlaylist(id: string, name: string): PlaylistDict {
    this.db.prepare('INSERT INTO playlists (id, name, created_at) VALUES (?, ?, ?)').run(id, name, new Date().toISOString());
    return this.getPlaylist(id)!;
  }

  updatePlaylistName(id: string, name: string): boolean {
    const r = this.db.prepare('UPDATE playlists SET name = ? WHERE id = ?').run(name, id);
    return r.changes > 0;
  }

  deletePlaylist(id: string): boolean {
    const r = this.db.prepare('DELETE FROM playlists WHERE id = ?').run(id);
    return r.changes > 0;
  }

  addSong(playlistId: string, song: { ytid: string; title: string; artist?: string; thumbnail?: string }): boolean {
    const existing = this.db.prepare('SELECT 1 FROM playlist_songs WHERE playlist_id = ? AND ytid = ?').get(playlistId, song.ytid);
    if (existing) return false;

    const maxOrder = this.db.prepare('SELECT MAX(song_order) FROM playlist_songs WHERE playlist_id = ?').get(playlistId) as { 'MAX(song_order)': number | null };
    const nextOrder = (maxOrder?.['MAX(song_order)'] ?? 0) + 1;

    this.db.prepare(
      'INSERT INTO playlist_songs (playlist_id, ytid, title, artist, thumbnail, song_order) VALUES (?, ?, ?, ?, ?, ?)'
    ).run(playlistId, song.ytid, song.title, song.artist ?? '', song.thumbnail ?? '', nextOrder);

    const pl = this.db.prepare('SELECT cover_thumbnail FROM playlists WHERE id = ?').get(playlistId) as { cover_thumbnail: string | null } | undefined;
    if (pl && !pl.cover_thumbnail && song.thumbnail) {
      this.db.prepare('UPDATE playlists SET cover_thumbnail = ? WHERE id = ?').run(song.thumbnail, playlistId);
    }
    return true;
  }

  removeSong(playlistId: string, ytid: string): boolean {
    const r = this.db.prepare('DELETE FROM playlist_songs WHERE playlist_id = ? AND ytid = ?').run(playlistId, ytid);
    if (r.changes > 0) {
      const songs = this.getPlaylistSongs(playlistId);
      const newCover = songs.length > 0 && songs[0].thumbnail ? songs[0].thumbnail : null;
      this.db.prepare('UPDATE playlists SET cover_thumbnail = ? WHERE id = ?').run(newCover, playlistId);
    }
    return r.changes > 0;
  }

  private toPlaylistDict(r: PlaylistRow): Omit<PlaylistDict, 'songs'> {
    return { id: r.id, name: r.name, coverThumbnail: r.cover_thumbnail, createdAt: r.created_at };
  }
}
